tls/s2n_client_hello.c

```diff
@@ -603,3 +603,54 @@ int s2n_client_hello_get_session_id(struct s2n_client_hello *ch, uint8_t *out, u
 
     return S2N_SUCCESS;
 }
+
+static S2N_RESULT s2n_client_hello_get_raw_extension(uint16_t extension_iana,
+        struct s2n_blob *raw_extensions, struct s2n_blob *extension)
+{
+    RESULT_ENSURE_REF(raw_extensions);
+    RESULT_ENSURE_REF(extension);
+
+    struct s2n_stuffer raw_extensions_stuffer = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&raw_extensions_stuffer, raw_extensions));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&raw_extensions_stuffer, raw_extensions->size));
+
+    while (s2n_stuffer_data_available(&raw_extensions_stuffer) > 0) {
+        uint16_t extension_type = 0;
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&raw_extensions_stuffer, &extension_type));
+
+        uint16_t extension_size = 0;
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&raw_extensions_stuffer, &extension_size));
+
+        uint8_t *extension_data = s2n_stuffer_raw_read(&raw_extensions_stuffer, extension_size);
+        RESULT_ENSURE_REF(extension_data);
+
+        if (extension_iana == extension_type) {
+            RESULT_GUARD_POSIX(s2n_blob_init(extension, extension_data, extension_size));
+            return S2N_RESULT_OK;
+        }
+    }
+    RESULT_BAIL(S2N_ERR_INVALID_ARGUMENT);
```

Comment:
Do we need this function to error if the extension isn't found? Can that information be found through whether or not the extension parameter is null or not?

Reply:
Good point, just have discuss regarding this with Lindsay as well. I believe error is not that necessary as we eventually going to ignore it. But not sure if check null pointer is the best practice. Notice that size may be 0, so only can check if data is null or not to tell exists or not. Any good suggestion we can follow here?

Reply:
I don't see a problem with the null check, do you @lrstewart?

Reply:
I updated the revision below, so s2n_client_hello_get_raw_extension return S2N_RESULT_OK if extension not exist.